### PR TITLE
:seedling: TK-3421 Add support to get specfic backup and backupplan

### PR DESCRIPTION
**Pull request by pankaj9310**

### CODE REVIEW

File: tools/target-browser/backup.go [hunk 1]
```diff
@@ -18,71 +19,104 @@ import (
...
 	if auth.UseHTTPS {
 		tvkURL.Scheme = internal.HTTPSscheme
 	}
-
-	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s?%s", tvkURL.String(), queryParam), nil)
+	var tvkURLPath string
+	if queryPath == "" {
+		tvkURLPath = fmt.Sprintf("%s?%s", tvkURL.String(), queryParam)
+	} else {
+		tvkURLPath = fmt.Sprintf("%s/%s?%s", tvkURL.String(), queryPath, queryParam)
```
Comment (msurudwad) [inline tools/target-browser/backup.go:58]: use path.Join

File: tools/target-browser/backup.go [hunk 2]
```diff
@@ -18,71 +19,104 @@ import (
...
 
 	tvkURL.Path = path.Join(tvkURL.Path, auth.TargetBrowserPath, apiPath)
 	tvkURL.Scheme = internal.HTTPscheme
 	if auth.UseHTTPS {
 		tvkURL.Scheme = internal.HTTPSscheme
 	}
-
-	req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("%s?%s", tvkURL.String(), queryParam), nil)
+	var tvkURLPath string
+	if queryPath == "" {
```
Comment (msurudwad) [inline tools/target-browser/backup.go:55]: rename to `pathParam`

File: tools/target-browser/backup.go [hunk 3]
```diff
@@ -18,71 +19,104 @@ import (
...
 	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return err
+		return "", err
 	}
 	if len(selector) > 0 {
+		if queryPath != "" {
+			body = parseData(body)
+		}
 		var backupBytes bytes.Buffer
```
Comment (msurudwad) [inline tools/target-browser/backup.go:85]: rename `backupBytes`

File: tools/target-browser/backup.go [hunk 4]
```diff
@@ -18,71 +19,104 @@ import (
...
-	return nil
+func parseData(respData []byte) []byte {
+	type result struct {
+		Result []interface{} `json:"results"`
+	}
+	var r result
+	r.Result = []interface{}{respData}
+	err := json.Unmarshal(respData, &r.Result[0])
+	if err != nil {
+		log.Fatal(err.Error())
```
Comment (msurudwad) [inline tools/target-browser/backup.go:115]: return err from here

File: tools/target-browser/backupplan.go [hunk 5]
```diff
@@ -13,11 +15,30 @@ type BackupPlanListOptions struct {
...
+			resp, err := auth.TriggerAPI(bpID, queryParam, apiPath, selector)
+			if err != nil {
+				return err
+			}
+			respData = append(respData, resp)
+		}
+		fmt.Println(respData)
+		return nil
+	}
+	resp, err := auth.TriggerAPI("", queryParam, apiPath, selector)
```
Comment (msurudwad) [inline tools/target-browser/backupplan.go:41]: check for err

File: tools/target-browser/backupplan.go [hunk 6]
```diff
@@ -13,11 +15,30 @@ type BackupPlanListOptions struct {
...
+	fmt.Println(resp)
+	return err
```
Comment (msurudwad) [inline tools/target-browser/backupplan.go:43]: return nil explicitly

File: tools/target-browser/backupplan.go [hunk 7]
```diff
@@ -13,11 +15,30 @@ type BackupPlanListOptions struct {
...
 	if err != nil {
 		return err
 	}
 	queryParam := values.Encode()
-	return auth.TriggerAPI(queryParam, internal.BackupPlanAPIPath, backupPlanSelector)
+	return auth.GetBackupOrBackupPlan(queryParam, internal.BackupPlanAPIPath, backupPlanSelector, backupPlanUIDs)
+}
+
+// GetBackupOrBackupPlan returns backup or backupPlan list stored on mounted target with available options
+func (auth *AuthInfo) GetBackupOrBackupPlan(queryParam, apiPath string, selector, args []string) error {
```
Comment (msurudwad) [inline tools/target-browser/backupplan.go:28]: rename this func

File: tools/target-browser/metadata.go [hunk 8]
```diff
@@ -19,5 +21,7 @@ func (auth *AuthInfo) GetMetadata(options *MetadataListOptions) error {
 		return err
 	}
 	queryParam := values.Encode()
-	return auth.TriggerAPI(queryParam, internal.MetadataAPIPath, []string{})
+	resp, apiErr := auth.TriggerAPI("", queryParam, internal.MetadataAPIPath, []string{})
+	fmt.Println(resp)
+	return apiErr
```
Comment (msurudwad) [inline tools/target-browser/metadata.go:26]: same here

**Review by msurudwad — CHANGES_REQUESTED** [review]

File: cmd/target-browser/cmd/backup.go [hunk 9]
```diff
@@ -20,11 +19,14 @@ func backupCmd() *cobra.Command {
 		Short:   "Get specific Backup or list of Backups",
 		Long: `Performs GET operation on target-browser's '/backup' API and gets specific Backup or list of Backups from mounted target location
 for specific backupPlan using available flags and options.`,
-		Example: `  # List of backups
+		Example: `  # List of backups for specific backupPlan
   kubectl tvk-target-browser get backup --backup-plan-uid <uid> --target-name <name> --target-namespace <namespace>
 
-  # Get specific backup (NOT SUPPORTED)
-  kubectl tvk-target-browser get backup --backup-plan-uid <uid> --backup-uid <uid> --target-name <name> --target-namespace <namespace>
+  # List of backups
```
[the remark this answers is not in the document]
Comment (pankaj9310) [reply cmd/target-browser/cmd/backup.go:25]: Updated docs on all places

File: tools/target-browser/backup.go (hunk 1 above)
Comment (pankaj9310) [reply tools/target-browser/backup.go]: Removed unused conditions

File: tools/target-browser/backup.go (hunk 2 above)
Comment (pankaj9310) [reply tools/target-browser/backup.go]: Renamed

File: tools/target-browser/backup.go (hunk 3 above)
Comment (pankaj9310) [reply tools/target-browser/backup.go]: renamed

File: tools/target-browser/backup.go (hunk 4 above)
Comment (pankaj9310) [reply tools/target-browser/backup.go]: returned err from here

File: tools/target-browser/backupplan.go (hunk 5 above)
Comment (pankaj9310) [reply tools/target-browser/backupplan.go:52]: Added condition to check error

File: tools/target-browser/backupplan.go (hunk 6 above)
Comment (pankaj9310) [reply tools/target-browser/backupplan.go]: Added condition to return error or nil

File: tools/target-browser/backupplan.go (hunk 7 above)
Comment (pankaj9310) [reply tools/target-browser/backupplan.go]: Rename function

File: tools/target-browser/metadata.go (hunk 8 above)
Comment (pankaj9310) [reply tools/target-browser/metadata.go]: done

File: tests/target-browser/target_browser_test.go [hunk 10]
```diff
@@ -415,12 +416,52 @@ var _ = Describe("Target Browser Tests", func() {
...
+			command := exec.Command(targetBrowserBinaryFilePath, args...)
+			output, err := command.CombinedOutput()
+			Expect(err).Should(HaveOccurred())
+			Expect(string(output)).Should(ContainSubstring("404 Not Found"))
+		})
+
+		It("Should get two backup for backup UID", func() {
+			args := []string{cmdGet, cmdBackup, backupUID, backupUID}
+			backupData := runCmdBackup(args)
+			Expect(len(backupData)).To(Equal(2))
```
Comment (msurudwad) [inline tests/target-browser/target_browser_test.go:456]: check for data as well like - whether same UIDs were returned which were requested.

File: tests/target-browser/target_browser_test.go [hunk 11]
```diff
@@ -415,12 +416,52 @@ var _ = Describe("Target Browser Tests", func() {
...
+		})
+
+		It("Should get one page backup without backupPlan UID", func() {
+			isLast = true
+			args := []string{cmdGet, cmdBackup, backupUID, flagPageSize, "1"}
+			backupData := runCmdBackup(args)
+			Expect(len(backupData)).To(Equal(1))
```
Comment (msurudwad) [inline tests/target-browser/target_browser_test.go:463]: same here

File: tests/target-browser/target_browser_test.go [hunk 12]
```diff
@@ -415,12 +416,52 @@ var _ = Describe("Target Browser Tests", func() {
...
+			args := []string{cmdGet, cmdBackupPlan, backupPlanUIDs[1]}
+			backupPlanData := runCmdBackupPlan(args)
+			Expect(len(backupPlanData)).To(Equal(1))
+			Expect(backupPlanData[0].BackupPlanUID).To(Equal(backupPlanUIDs[1]))
+		})
+
+		It("Should get two backupPlan for backupPlan UIDs", func() {
+			args := []string{cmdGet, cmdBackupPlan, backupPlanUIDs[0], backupPlanUIDs[1]}
+			backupPlanData := runCmdBackupPlan(args)
+			Expect(len(backupPlanData)).To(Equal(2))
```
Comment (msurudwad) [inline tests/target-browser/target_browser_test.go:434]: check for data as well like - whether same UIDs were returned which were requested.

**Review by msurudwad — COMMENTED** [review]
LGTM

File: tests/target-browser/target_browser_test.go (hunk 10 above)
Comment (pankaj9310) [reply tests/target-browser/target_browser_test.go:458]: done

File: tests/target-browser/target_browser_test.go (hunk 12 above)
Comment (pankaj9310) [reply tests/target-browser/target_browser_test.go:434]: done

**Review by msurudwad — APPROVED** [review]
LGTM

**Review by madhurtrilio — APPROVED** [review]

**Pull request merged**